topdown/eval_test.go

```diff
@@ -443,3 +445,603 @@ func TestTopdownVirtualCache(t *testing.T) {
+			exp:   `[{"x": 1}]`,
+		},
+		{
+			note: "deep query into partial object, on first var in ref, multiple vars",
+			module: `package test
+				p.q[r][s] := 1 { r := "foo"; s := "bar" }
+			`,
+			query: `data.test.p.q.foo = x`,
+			exp:   `[{"x": {"bar": 1}}]`,
+		},
+		{
+			note: "deep query into partial object, beyond first var in ref, multiple vars",
+			module: `package test
+				p.q[r][s] := 1 { r := "foo"; s := "bar" }
+			`,
+			query: `data.test.p.q.foo.bar = x`,
+			exp:   `[{"x": 1}]`,
+		},
+		{
+			note: "deep query into partial object, beyond first var in ref, multiple vars",
+			module: `package test
+				p.q[r][s].t := 1 { r := "foo"; s := "bar" }
+			`,
+			query: `data.test.p.q.foo.bar = x`,
+			exp:   `[{"x": {"t": 1}}]`,
+		},
+		{
+			note: "deep query to partial object, overlapping rules (key override), no dynamic ref",
+			module: `package test
+				p.q[r] := 1 { r := "foo" }
+				p.q.r := 2
+			`,
+			query: `data.test.p.q = x`,
+			exp:   `[{"x": {"foo": 1, "r": 2}}]`,
+		},
+		{
+			note: "deep query into partial object, overlapping rules (key override), no dynamic ref",
+			module: `package test
+				p.q[r] := 1 { r := "foo" }
+				p.q.r := 2
+			`,
+			query: `data.test.p.q.r = x`,
+			exp:   `[{"x": 2}]`,
+		},
+		{
+			note: "deep query into partial object, overlapping rules, no dynamic ref",
+			module: `package test
+				p.q[r] := 1 { r := "foo" }
+				p.q[r] := 2 { r := "bar" }
+			`,
+			query: `data.test.p.q.foo = x`,
+			exp:   `[{"x": 1}]`,
+		},
+		{
+			note: "deep query into partial object, overlapping rules with same key/value, no dynamic ref",
+			module: `package test
+				p.q[r] := 1 { r := "foo" }
+				p.q[r] := 1 { r := "foo" }
+			`,
+			query: `data.test.p.q.foo = x`,
+			exp:   `[{"x": 1}]`,
+		},
+		{
+			note: "deep query into partial object, overlapping rules, dynamic ref",
+			module: `package test
+				p.q[r].s := 1 { r := "r" }
+				p.q.r[s] := 2 { s := "foo" }
+			`,
+			query: `data.test.p.q.r = x`,
+			exp:   `[{"x": {"s": 1, "foo": 2}}]`,
+		},
+		{
+			note: "deep query into partial object, overlapping rules with same key/value, dynamic ref",
+			module: `package test
+				p.q[r].s := 1 { r := "r" }
+				p.q.r[s] := 1 { s := "s" }
+			`,
+			query: `data.test.p.q.r = x`,
+			exp:   `[{"x": {"s": 1}}]`,
+		},
+		// Multiple results
+		//{
+		//	note: "deep query to partial object, overlapping rules with same key/value, dynamic ref",
+		//	module: `package test
+		//		p.q[r] := 1 { r := "foo" }
+		//		p.q[r] := 2 { r := "bar" }
+		//	`,
+		//	query: `i := ["foo", "bar"][_]; data.test.p.q[i] = x`,
+		//	exp:   `[{"x": {"s": 1}}]`,
+		//},
+		//{
+		//	note: "deep query into partial object, overlapping rules with same key/value, dynamic ref",
+		//	module: `package test
+		//		p.q[r].s := 1 { r := "foo" }
+		//		p.q[r].s := 2 { r := "bar" }
+		//	`,
+		//	query: `i := ["foo", "bar"][_]; data.test.p.q[i].s = x`,
+		//	exp:   `[{"x": 1}, {"x": 2}]`,
+		//},
```

Comment:
What's up with these? Do they not work yet; or have they been subsumed by the tests above...? 🤔

Reply:
Ah, I missed un-commenting those. Fixed 👍 .